README.md
<p align="center">
  <a href="https://kinopoisk.dev/" target="blank"><img src="https://openmovieapi.dev/full-size-cover.png" alt="Open movie API logo" /></a>
</p>

# TS/JS клиент для kinopoisk.dev API (@openmoviedb/kinopoiskdev_client)

Эта библиотека позволяет взаимодействовать с [неофициальным API kinopoisk](kinopoisk.dev), обеспечивая удобный доступ к таким сущностям как:

- Фильмы (movie)
- Сезоны (season)
- Персоны (person)
- Ключевые слова (keyword)
- Студии (studio)
- Изображения (image)
- Отзывы (review)

## Начало работы

1. Получите токен у телеграм бота [@kinopoiskdev_bot](https://t.me/kinopoiskdev_bot).
2. Изучите [документацию](https://api.kinopoisk.dev/v1/documentation). Она выполнена в формате OpenAPI и поможет вам понять, какие данные можно получить и какие параметры можно передать в запросе.

## Установка

Установите библиотеку через npm:

```bash
npm i @openmoviedb/kinopoiskdev_client
```

или yarn:

```bash
yarn add @openmoviedb/kinopoiskdev_client
```

## Использование

Сначала инициализируйте клиент и передайте в него ваш токен:

```typescript
import { KinopoiskDev } from '@openmoviedb/kinopoiskdev_client';

const kp = new KinopoiskDev('ВАШ ТОКЕН');
```

### Взаимодействие с сущностями

После инициализации клиента вы можете начать взаимодействовать с различными сущностями. Вот пример того, как можно получить информацию о фильме по его ID:

```typescript
const {data} = await kp.movie.getById(666);
console.log(data);
```

### Использование Query Builder

QueryBuilder это специальный инструмент, который помогает вам создавать сложные запросы без написания их вручную. QueryBuilder позволяет вам добавлять различные параметры в ваш запрос, такие как сортировка, фильтрация, пагинация и выбор конкретных полей для возвращения. Этот инструмент валидирует поля, которые вы передаете, и подсвечивает подсказки, что облегчает использование. 

Вот пример использования QueryBuilder для получения списка фильмов с определенными параметрами:

```typescript
import {
  KinopoiskDev,
  MovieQueryBuilder,
  SPECIAL_VALUE,
  SORT_TYPE,
} from '@openmoviedb/kinopoiskdev_client';

const kp = new KinopoiskDev('ВАШ ТОКЕН');

const getRelatedByQueryBuilderMovies = async () => {
  const queryBuilder = new MovieQueryBuilder();

  const query = queryBuilder
    .select(['id', 'name', 'rating', 'poster', 'year'])
    .filterRange('year', [2020, 2023])
    .filterRange('rating.kp', [7.5, 10])
    .filterExact('poster.url', SPECIAL_VALUE.NOT_NULL)
    .sort('rating.kp', SORT_TYPE.DESC)
    .paginate(1, 10)
    .build();

  const { data, error, message } = await kp.movie.search(query);

  console.log(data, error, message);
};

getRelatedByQueryBuilderMovies();
```

Обратите внимание, что QueryBuilder совместим с axios и другими HTTP клиентами. Если вам нужно использовать другой HTTP клиент, вы можете сделать это. Библиотека предоставляет все необходимые типы для этого.

Также вы можете взаимодействовать с API без использования QueryBuilder, просто составив запрос самостоятельно или в интерфейсе документации API. Пример:

```typescript
import { KinopoiskDev, Filter, MovieFields } from '@openmoviedb/kinopoiskdev_client';

const kp = new KinopoiskDev('ВАШ ТОКЕН');

const getRelatedWithoutQueryBuilderMovies = async () => {
  const query: Filter<MovieFields> = {
    selectFields: ['id', 'name', 'rating', 'poster', 'year'],
    year: '2020-2023',
    'rating.kp': '7.5-10',
    'poster.url': '!null',
    sortField: 'rating.kp',
    sortType: '-1',
    page: 1,
    limit: 10,
  };

  const { data, error, message } = await kp.movie.getByFilters(query);

  console.log(data, error, message);
};

getRelatedWithoutQueryBuilderMovies();
```

### Подробные примеры
Для более детальных примеров использования библиотеки, пожалуйста, обратитесь к директории [examples](https://github.com/OpenMovieDB/kinopoiskdev_client/tree/main/examples) в репозитории проекта. Там вы найдете примеры работы с каждой из доступных сущностей.


CHANGELOG.md
# Changelog

## [1.4.0](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.10...v1.4.0) (2023-11-18)


### Features

* add v1.4 to builder ([1d0c5bf](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/1d0c5bf9b5bfa2f7973af7de7cb00e4825d0c199))
* update images methods to v1.4 ([7e03e46](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/7e03e46ab60b330d4a08377e088674bdef89c927))

## [1.3.10](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.9...v1.3.10) (2023-07-22)


### Bug Fixes

* update build ([a217822](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/a217822d6688a9c131d4ce1124e5466544881dd9))

## [1.3.9](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.8...v1.3.9) (2023-07-22)


### Bug Fixes

* **query.builder:** incorrect field types are used ([ee31ef1](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/ee31ef1a0a8d1dfa86b6c4de6a81903a268c91df))

## [1.3.8](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.7...v1.3.8) (2023-07-04)


### Bug Fixes

* update package description to be more concise and accurate ([1988367](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/19883673211d599e7aaa1209d4d510a09d5c99cf))

## [1.3.7](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.6...v1.3.7) (2023-06-26)


### Bug Fixes

* invalid url in getPossibleValuesByField ([6b35da3](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/6b35da32ef22cac401f3c5b60229dd22d611fa05))
* search methods do not use pagination ([b989bd9](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/b989bd98f06126f0b0a26813fb8921c42e030b1d))

## [1.3.6](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.5...v1.3.6) (2023-06-25)


### Bug Fixes

* bundle is too big ([0196cfc](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/0196cfcd0fdcc944cdf2be12d59839949297af3a))

## [1.3.5](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.4...v1.3.5) (2023-06-25)


### Bug Fixes

* remove unused files on npm ([423b034](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/423b03472fd51be00547231827a17d81e35c60f1))
* update version ([dd73df5](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/dd73df5cb1485fc870e58ed29d8855f17cba390e))

## [1.3.4](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.3...v1.3.4) (2023-06-25)


### Bug Fixes

* commonjs causing conflicts ([5e41b02](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/5e41b02e8247e5c5c22cc4a5441162d860863ffa))

## [1.3.3](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.2...v1.3.3) (2023-06-25)


### Bug Fixes

* commonjs causing conflicts ([96e740a](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/96e740ac8fc9b88d68be38c9a1fb7f5522d84e63))

## [1.3.2](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.1...v1.3.2) (2023-06-25)


### Bug Fixes

* commonjs causing conflicts ([18f9f90](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/18f9f90bc49fb6dbde9c480b5c9d2345b583088a))

## [1.3.1](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.3.0...v1.3.1) (2023-06-25)


### Bug Fixes

* commonjs causing conflicts ([ad24128](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/ad24128aa04d39153f8db31d7fca93fc265c30d7))

## [1.3.0](https://github.com/OpenMovieDB/kinopoiskdev_client/compare/v1.2.0...v1.3.0) (2023-06-25)


### Features

* add .idea and package.json files to the project ([8bdcb07](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/8bdcb07af5a5f874e1deb2caa274a9800ac37e93))
* add default export class KinopoiskDev to index.ts ([16f9572](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/16f95726aa48b647353a06eb04666d04fab10303))
* add release-please configuration file and package to enable automatic releases ([c1a381e](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/c1a381e8fb9fb17af9348a0f68ae2228518a1f2a))
* **api.interface.ts:** add new file for API interfaces ([01f7b7f](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/01f7b7f80c18c810c2946159f95441d898de7c54))
* **client-request.ts:** add ClientRequest class to handle API requests ([2d5feda](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/2d5fedaa56c4b6c3fa6de80ebb7c067d6cdd3018))
* **client-request.ts:** add error handling to get method and create IResponseError interface ([b91c057](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/b91c0570d4d2ba8fb7ef468c9ce05115d4e4cc05))
* **client-request.ts:** add IResponse interface to handle response data and status code ([031c604](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/031c604fb707b11ad94bc0162e4b6b0e0272c82f))
* **client-request.ts:** add queryParams method to handle query parameters in GET requests ([d43b8e2](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/d43b8e22c003b774e6e4a55dd950679ac7ce5ccf))
* **client-request.ts:** add version parameter to get method and use URLSearchParams for query parameters ([ebf90df](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/ebf90df31c0df603681a9c45944974db90168c4a))
* **client-request:** add error handling for failed requests in get method ([7879f4b](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/7879f4bd1ca02240c7601aeea50220d424a25c94))
* **client-request:** add support for query parameters in get method ([7879f4b](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/7879f4bd1ca02240c7601aeea50220d424a25c94))
* **enums/version.enum.ts:** add VERSIONS enum ([ebf90df](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/ebf90df31c0df603681a9c45944974db90168c4a))
* **examples:** add example package.json and movie.ts script with usage of kinopoiskdev_ts_client API ([91ad1c1](https://github.com/OpenMovieDB/kinopoiskdev_client/commit/91ad1c1c5edc26ca60b050a1654bc42ee939bd95))

examples/src/movie.ts
import {
  KinopoiskDev,
  MovieQueryBuilder,
  SPECIAL_VALUE,
  SortType,
  Filter,
  MovieFields,
} from '@openmoviedb/kinopoiskdev_client';

const kp = new KinopoiskDev('ВАШ ТОКЕН');

// Получение списка фильмов за 2020-2023 годы с рейтингом от 7.5 до 10, у которых есть постер и которые созданы в России или США
// Этот пример использует query builder
const getRelatedByQueryBuilderMovies = async () => {
  // Создаем билдер запросов для фильмов
  const queryBuilder = new MovieQueryBuilder();

  // Выбираем поля, которые мы хотим получить в ответе
  // Полный список полей можно посмотреть в документации
  // https://api.kinopoisk.dev/v1/documentation для метода /v1.3/movie
  const query = queryBuilder
    .select(['id', 'name', 'rating', 'poster', 'year'])
    // Добавляем фильтр поиска по указанному диапазону года
    .filterRange('year', [2020, 2023])
    // Добавляем фильтр поиска по указанному диапазону рейтинга
    .filterRange('rating.kp', [7.5, 10])
    // Добавляем фильтр для поиска фильмов с постером
    .filterExact('poster.url', SPECIAL_VALUE.NOT_NULL)
    // Добавим страны
    .filterExact('countries.name', 'США')
    .filterExact('countries.name', 'Россия')
    // Добавляем сортировку по рейтингу
    .sort('rating.kp', SortType.DESC)
    // Добавляем пагинацию и получаем 1 страницу по с 10 фильмами на странице
    .paginate(1, 10)
    // Собираем запрос
    .build();

  // Отправляем запрос на получение фильмов
  const { data, error, message } = await kp.movie.getByFilters(query);

  if (data) {
    const { docs, page, limit } = data;
    console.log(`Страница ${page} из ${limit}`);
    console.log(docs);
  }

  // Если будет ошибка, то выведем ее в консоль
  if (error) console.log(error, message);
};

// Этот пример без использования query builder. Объект для запроса создается вручную на основе документации API
const getRelatedWithoutQueryBuilderMovies = async () => {
  const query: Filter<MovieFields> = {
    // Выбираем поля, которые мы хотим получить в ответе
    selectFields: ['id', 'name', 'rating', 'poster', 'year'],
    // Добавляем фильтр поиска по указанному диапазону года
    year: '2020-2023',
    // Добавляем фильтр поиска по указанному диапазону рейтинга
    'rating.kp': '7.5-10',
    // Добавляем фильтр для поиска фильмов с постером
    'poster.url': '!null',
    // Добавляем сортировку по рейтингу
    sortField: 'rating.kp',
    sortType: '-1',
    // Добавляем пагинацию и получаем 1 страницу по с 10 фильмами на странице
    page: 1,
    limit: 10,
  };

  // Отправляем запрос на получение фильмов
  const { data, error, message } = await kp.movie.getByFilters(query);

  if (data) {
    const { docs, page, limit } = data;
    console.log(`Страница ${page} из ${limit}`);
    console.log(docs);
  }

  // Если будет ошибка, то выведем ее в консоль
  if (error) console.log(error, message);
};

// Получение фильма по id
const getMovieById = async () => {
  const { data, error, message } = await kp.movie.getById(666);

  if (data) {
    console.log(data);
  }

  // Если будет ошибка, то выведем ее в консоль
  if (error) console.log(error, message);
};

// Получить рандомный фильм с использованием стандартного метода
// Вернутся все поля фильма как в методе getById
const getRandomMovie = async () => {
  const { data, error, message } = await kp.movie.getRandom();

  if (data) {
    console.log(data);
  }

  // Если будет ошибка, то выведем ее в консоль
  if (error) console.log(error, message);
};

// Получить рандомный фильм с указанными фильтрами
// Вернутся только указанные поля фильма или поля по умолчанию, если ничего не указано
const getRandomMovieWithFilters = async () => {
  // Создаем билдер запросов для фильмов
  const queryBuilder = new MovieQueryBuilder();

  // Выбираем поля, которые мы хотим получить в ответе
  const baseQuery = queryBuilder
    .select(['id', 'name', 'rating', 'poster', 'year'])
    // Добавляем фильтр для поиска фильмов с постером
    .filterExact('poster.url', SPECIAL_VALUE.NOT_NULL)
    // Добавляем фильтр для поиска фильмов с трейлерами
    .filterExact('videos.trailers.url', SPECIAL_VALUE.NOT_NULL);

  const firstQuery = baseQuery
    // получим первую страницу, чтобы узнать сколько всего фильмов удовлетворяет нашим фильтрам
    .paginate(1, 1)
    // Собираем запрос
    .build();

  // Отправляем запрос на получение количества фильмов
  const firstRes = await kp.movie.getByFilters(firstQuery);
  if (firstRes.data) {
    const { pages } = firstRes.data;
    // Генерируем случайное число от 1 до pages
    const randomPage = Math.floor(Math.random() * pages) + 1;
    const query = baseQuery.paginate(randomPage, 1).build();

    const { data, error, message } = await kp.movie.getByFilters(query);

    if (data) {
      console.log(data.docs[0]);
    }
    // Если будет ошибка, то выведем ее в консоль
    if (error) console.log(error, message);
  }

  // Если будет ошибка, то выведем ее в консоль
  if (firstRes.error) console.log(firstRes.error, firstRes.message);
};

// Найти фильмы по названию
const searchMovies = async () => {
  const queryBuilder = new MovieQueryBuilder();
  // Создаем запрос для поиска фильмов по подходящих под наш запрос
  const query = queryBuilder
    // Указываем что хотим получить фильм под названием Аватар вышедший в 2022
    .query('Аватар 2022')
    // Добавляем пагинацию и получаем 1 страницу по с 10 фильмами на странице
    .paginate(1, 10)
    // Собираем запрос
    .build();

  const { data, error, message } = await kp.movie.getBySearchQuery(query);

  if (data) {
    const { docs, page, limit } = data;
    console.log(`Страница ${page} из ${limit}`);
    console.log(docs);
  }

  // Если будет ошибка, то выведем ее в консоль
  if (error) console.log(error, message);
};

// Получить все возможные жанры
const getGenres = async () => {
  // Отправляем запрос на получение жанров
  const { data, error, message } = await kp.movie.getPossibleValuesByField(
    'genres.name',
  );

  if (data) console.log(data);

  // Если будет ошибка, то выведем ее в консоль
  if (error) console.log(error, message);
};

// Получить все возможные страны
const getCountries = async () => {
  // Отправляем запрос на получение стран
  const { data, error, message } = await kp.movie.getPossibleValuesByField(
    'countries.name',
  );

  if (data) console.log(data);

  // Если будет ошибка, то выведем ее в консоль
  if (error) console.log(error, message);
};

const bootstrap = async () => {
  await getMovieById();
  await getRelatedByQueryBuilderMovies();
  await getRelatedWithoutQueryBuilderMovies();
  await getRandomMovie();
  await getRandomMovieWithFilters();
  await searchMovies();
  await getGenres();
  await getCountries();
};

bootstrap();
